### ISSUE TITLE
Compiling ISPC on Xeon Phi for an Intel-sponsored CMU course assignment

### ISSUE THREAD

**Issue by harshavardhanpandit** [ispc/ispc#1103]
Hi,

I am Harsha, a graduate student at CMU - I am currently working with Prof. Kayvon Fatahalian (cc'ed) on an Intel-sponsored assignment on the Xeon Phi for the course "15-418: Parallel Computer Architecture and Programming".

Our goal here is to be able to compile, offload and run ISPC programs on the Xeon Phi. I am, however, facing some issues with this build process and was hoping I would receive some help in this direction.

[ISPC Xeon Phi Example](https://github.com/harshavardhanpandit/ISPC_XeonPhi/tree/master/ISPC_XeonPhi_Example) - this is a link containing a small example of how I am currently building and running ISPC programs on the Phi. It contains a Makefile, an ISPC file, a cpp wrapper for the ISPC file and a offload.cpp file which contains the main routine. (I have also pasted the "knc.h" in the folder so that this application directly builds by just typing "make") When I make this application, it currently gives me the following warning:

dispatch.cpp(6): warning #2570: *MIC* function has not been declared with compatible "target" attribute
    ispc::foo(n, data);

The binary does run on the Xeon Phi, but I was wondering how I could avoid these warnings. The header file auto-generated by ISPC does not have the target attribute declaration. Could you please let me know how I can make this work seamlessly, and/or what is the correct way to build and run such programs?

The logical extension of the previous question is about how I would compile an ISPC code which launches tasks on the Xeon Phi. Could you please also let me know what is the correct way to compile such ISPC files along with the "tasksys.cpp"?

I would much appreciate any inputs and help that address these issues.

Thanks,
Harsha

**Comment by egaburov** [comment]
Hi,

My understanding is that the warning is due to the ISPC generated offload_ispc.h  file doesn’t have 
	__declspec(target(mic))
before
    extern void foo(const int32_t n, float * data);

Changing the line in offload_ispc.h to
    __declspec(target(mic)) extern void foo(const int32_t n, float * data);
fixes the warning for the following compilation line:
    icc offload_ispc.o  dispatch.cpp offload.cpp -o offload 	-offload=optional

ISPC doesn’t generate this line automatically in .h file, ergo the warning. 

One workaround would be to annotate export functions with _mic postfix, for example, and parse .h files to add __declspec(target(mic)) in front of each function definition with _mic postfix, before building the source with icc.

I don’t know whether ISPC will add an option to include __declspec(target(mic)) attribute in .h files for export functions, to enable MIC offloading.

Cheers,
  Evghenii

**Comment by egaburov** [comment]
Missed the last part:
> The logical extension of the previous question is about how I would compile an ISPC code which launches tasks on the Xeon Phi. Could you please also let me know what is the correct way to compile such ISPC files along with the "tasksys.cpp"?
> 
> 
Please see the following: https://github.com/ispc/ispc/tree/master/examples/portable <https://github.com/ispc/ispc/tree/master/examples/portable>

This one contains parallel portable examples, that build and run on CPU, XeonPhi and Kepler GPU (for overview see a poster presented at GTC last year http://on-demand.gputechconf.com/gtc/2014/poster/pdf/P4118_compilers_XeonPhi_GPU_ISPC.pdf <http://on-demand.gputechconf.com/gtc/2014/poster/pdf/P4118_compilers_XeonPhi_GPU_ISPC.pdf>)
 
Each example has a makefile per target: Makefile_cpu for CPU, Makefile_knc for XeonPhi, and Makefile_ptx for Kepler GPU, and all use the same .ispc kernels.

In the examples, the execution on MIC is native, but it shouldn’t be difficult to offload this with the offload example you have: the export functions uses only 1 thread and launches its own tasks with the help of tasksys.cpp and OpenMP.

The makefiles are tailored for ICC compilers, except the Makefile_ptx which uses g++.

Cheers,
  Evghenii

**Comment by dbabokin** [comment]
Indeed, we are not adding target(mic) declspec/attribute, as we are targeting native, not offload model. But there's not big difference.

We may add "--offload" switch to add a it to the generated file when required.

As a workaround, instead of adding declspec to each function you can mark the whole region with
#pragma offload_attribute(push, target(mic))
#pragma offload_attribute(pop)

**Comment by harshavardhanpandit** [comment]
@egaburov, @dbabokin - Thanks so much, that helps a lot !

**Comment by harshavardhanpandit** [comment]
@egaburov - I tried to combine the offloading and compiling with tasksys.cpp and OpenMP as suggested in the link you provided but am running into some problems. 

Here is the link to the offload example with tasks - [ISPC Offload with Tasks](https://github.com/harshavardhanpandit/ISPC_XeonPhi/tree/master/ISPC_offload_with_tasks)
I have added a simple parser to add the pragmas in the .h file and now the example can be compiled by just running ./compile.sh (instead of "make"). But I think I am definitely going wrong in compiling and linking the tasksys.cpp since I get the following errors:

`offload_ispcMIC.o: In function foo_with_task___Cuniun_3C_unf_3E_:`
`offload_ispcMIC.cpp:(.text+0x196): undefined reference to ISPCAlloc`
`offload_ispcMIC.cpp:(.text+0x1cb): undefined reference to ISPCLaunch`
`offload_ispcMIC.cpp:(.text+0x1d9): undefined reference to ISPCSync`
`offload_ispcMIC.o: In function foo_with_task:`
`offload_ispcMIC.cpp:(.text+0x2bf): undefined reference to ISPCAlloc`
`offload_ispcMIC.cpp:(.text+0x2f8): undefined reference to ISPCLaunch`
`offload_ispcMIC.cpp:(.text+0x306): undefined reference to ISPCSync`

Could you please help me figure out what's going wrong?

Thanks!
Harsha

**Comment by jbrodman** [comment]
Did you modify the tasksys.cpp functions to also be offloadable?

**Comment by harshavardhanpandit** [comment]
@jbrodman - Ohh right, I haven't. Thanks, I'll try that.

**Comment by aneshlya** [comment]
This issue seems to be resolved. I'm closing it.
